File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Delete_the_Middle_Node_of_a_Linked_List/inseo.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur = head
        n = 0

        while cur != None:
            n += 1
            cur = cur.next
        
        if n==1:
            return None

        cur_idx = 0
        cur = head

        while cur != None:
            if cur_idx == int(n/2)-1:
                cur.next = cur.next.next if cur.next != None else None
                break
            else:
                cur = cur.next
                cur_idx += 1
        
        return head

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Delete_the_Middle_Node_of_a_Linked_List/inseo.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        if(head==NULL )
            return head;
        if(head->next==NULL)
            return nullptr;
        int s=0;
        ListNode* temp =head;
        while(temp!=NULL){
            s++;
            temp=temp->next;
        }
        int n=s/2;
        int i=1;
        
        temp=head;
        while(temp!=NULL){
            if(i==n){
                temp->next=temp->next->next;
                break;
            }
            temp=temp->next;
            i++;
        }
        return head;
    }
};